crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,753 @@
+/// In order to initialize the caching layer, the path to the cache file is provided along with the
+/// max number of items that are stored in the in-memory LRU cache.
+///
+/// One can load the cache from the file system by calling `load_cache` and save the cache to the
+/// file system by calling `save_cache`.
+#[derive(Debug, Clone)]
+pub struct CacheLayer {
+    /// In-memory LRU cache, mapping requests to responses.
+    cache: SharedCache,
+}
+
+impl CacheLayer {
+    /// Instantiate a new cache layer with the the maximum number of
+    /// items to store.
+    pub fn new(max_items: u32) -> Self {
+        Self { cache: SharedCache::new(max_items) }
+    }
+
+    /// Returns the maximum number of items that can be stored in the cache, set at initialization.
+    pub const fn max_items(&self) -> u32 {
+        self.cache.max_items()
+    }
+
+    /// Returns the shared cache.
+    pub fn cache(&self) -> SharedCache {
+        self.cache.clone()
+    }
+}
+
+impl<P, T, N> ProviderLayer<P, T, N> for CacheLayer
+where
+    P: Provider<T, N>,
+    T: Transport + Clone,
+    N: Network,
+{
+    type Provider = CacheProvider<P, T, N>;
+
+    fn layer(&self, inner: P) -> Self::Provider {
+        CacheProvider::new(inner, self.cache())
+    }
+}
+
+/// The [`CacheProvider`] holds the underlying in-memory LRU cache and overrides methods from the
+/// [`Provider`] trait that should attempt to fetch from cache and fallback to the RPC in case of a
+/// cache miss.
+///
+/// Most importantly, the [`CacheProvider`] adds `save_cache` and `load_cache` methods to the
+/// provider interface to lets users save cache to the disk and load from it on demand.
+#[derive(Debug, Clone)]
+pub struct CacheProvider<P, T, N> {
+    /// Inner provider.
+    inner: P,
+    /// In-memory LRU cache, mapping requests to responses.
+    cache: SharedCache,
+    /// Phantom data
+    _pd: PhantomData<(T, N)>,
+}
+
+impl<P, T, N> CacheProvider<P, T, N>
+where
+    P: Provider<T, N>,
+    T: Transport + Clone,
+    N: Network,
+{
+    /// Instantiate a new cache provider.
+    pub const fn new(inner: P, cache: SharedCache) -> Self {
+        Self { inner, cache, _pd: PhantomData }
+    }
+}
+
+/// Attempts to fetch the response from the cache by using the hash of the request params.
+///
+/// In case of a cache miss, fetches from the RPC and saves the response to the cache.
+///
+/// This helps overriding [`Provider`] methods that return [`TransportResult<T>`].
+macro_rules! cache_get_or_fetch {
+    ($self:expr, $req:expr, $fetch_fn:expr) => {{
+        let hash = $req.params_hash()?;
+        if let Some(cached) = $self.cache.get_deserialized(&hash)? {
+            return Ok(Some(cached));
+        }
+
+        let result = $fetch_fn.await?;
+        if let Some(ref data) = result {
+            let json_str = serde_json::to_string(data).map_err(TransportErrorKind::custom)?;
+            let _ = $self.cache.put(hash, json_str)?;
+        }
+
+        Ok(result)
+    }};
+}
```

Comment:
I'm a bit sceptical that we even need these macros if those could also just be functions, but this also works, so lets stick with that